lib/Sema/CSBindings.cpp

```diff
@@ -481,8 +481,173 @@ void PotentialBindings::finalize(
   }
 }
 
-PotentialBindings::BindingScore
-PotentialBindings::formBindingScore(const PotentialBindings &b) {
+void BindingSet::addBinding(PotentialBinding binding) {
+  if (Bindings.count(binding))
+    return;
+
+  // If this is a non-defaulted supertype binding,
+  // check whether we can combine it with another
+  // supertype binding by computing the 'join' of the types.
+  if (binding.isViableForJoin()) {
+    auto isAcceptableJoin = [](Type type) {
+      return !type->isAny() && (!type->getOptionalObjectType() ||
+                                !type->getOptionalObjectType()->isAny());
+    };
+
+    SmallVector<PotentialBinding, 4> joined;
+    for (auto existingBinding = Bindings.begin();
+         existingBinding != Bindings.end();) {
+      if (existingBinding->isViableForJoin()) {
+        auto join =
+            Type::join(existingBinding->BindingType, binding.BindingType);
+
+        if (join && isAcceptableJoin(*join)) {
+          // Result of the join has to use new binding because it refers
+          // to the constraint that triggered the join that replaced the
+          // existing binding.
+          joined.push_back(binding.withType(*join));
+          // Remove existing binding from the set.
+          // It has to be re-introduced later, since its type has been changed.
+          existingBinding = Bindings.erase(existingBinding);
+          continue;
+        }
+      }
+
+      ++existingBinding;
+    }
+
+    for (const auto &binding : joined)
+      (void)Bindings.insert(binding);
+
+    // If new binding has been joined with at least one of existing
+    // bindings, there is no reason to include it into the set.
+    if (!joined.empty())
+      return;
+  }
+
+  // Check whether the given binding covers any of the literal protocols
+  // associated with this type variable.
+  {
+    bool allowsNil = canBeNil();
+
+    for (auto &literal : Literals) {
+      auto *protocol = literal.first;
+
+      // Skip conformance to `nil` protocol since it doesn't
+      // have a default type and can't affect binding set.
+      if (protocol->isSpecificProtocol(
+              KnownProtocolKind::ExpressibleByNilLiteral))
+        continue;
+
+      auto &info = literal.second;
+
+      if (!info.viableAsBinding())
+        continue;
+
+      bool isCovered = false;
+      Type adjustedTy;
+
+      std::tie(isCovered, adjustedTy) =
+          isLiteralCoveredBy(info, binding, allowsNil);
+
+      if (!isCovered)
+        continue;
+
+      binding = binding.withType(adjustedTy);
+      info.setCoveredBy(binding.getSource());
+    }
+  }
+
+  Bindings.insert(std::move(binding));
+>>>>>>> [CSBindings] Separate inference from final product
```

Comment:
Looks like a git conflict marker left in this commit?

Reply:
Ah, it might be old code which I fixed :)